doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,428 @@
+# Platform V2: Txn Management
+
+## Context
+
+Our SQL Table Transactions Management still shows some of its lineage from the
+single-binary. At times, this conflicts with the "Platform V2" goals of
+_use-case isolation_ and _zero-downtime upgrades (0dt)_.
+
+## Goals
+
+Use-case isolation. Make it possible to prevent pgwire sessions from interfering
+with each other through:
+
+- Large reads
+- Large writes
+- Frequent reads
+- Frequent writes
+- Long-running transactions
+
+Zero-downtime upgrades. Nothing in particular motivating here. Don't introduce
+anything that interferes with 0dt.
+
+Don't break things that work today:
+- Explicit SQL transactions that are entirely reads or entirely writes (plus
+  "constant" reads)
+- Multi-table read transactions
+- Single-table write transactions
+- Reads include from `MATERIALIZED VIEW`s
+- Transactions that are implicitly reads followed by writes:
+  - `INSERT INTO ... SELECT FROM ...`
+  - `UPDATE`
+  - `DELETE`
+- Don't regress performance "too much". Moving to a distributed system is
+  necessary to provide isolation and distributed transactions are necessarily
+  less performant.
+
+Nice to haves, while we're revisiting SQL tables and transactions:
+- A path to multi-table write transactions. It need not end at the global
+  optimum, just be possible.
+- Bounded memory usage for large transactions
+- Highly optimized `COPY INTO`/`COPY FROM`
+
+## Non-Goals
+
+- Preventing pgwire session disconnection during upgrades and other restarts.
+  Node failures and loadbalancer termination are rare but impossible to
+  eliminate, so users will always need a connection retry loop.
+- Performance under highly-contended writes. Materialize will never be as good
+  at this as a dedicated OLTP database.
+- Transactions that arbitrarily interleave reads and writes. This could
+  potentially be done for special cases, but the general case of arbitrary
+  SELECTs requires a way to include uncommitted data in compute dataflow inputs.
+  This could potentially be modeled as a compound timestamp, but that's an
+  immense change.
+- Align table and source upper advancements. This is a lovely thing to do, but
+  seems independent of this work.
+
+## Overview
+
+Each table is (and will continue to be) backed by an individual persist _shard_
+with a unique _ShardId_ address.
+
+Currently:
+- The data in each of these shards is written already _reclocked_ (roughly, with
+  a timestamp suitable for use directly in compute dataflows). One of the
+  seriously considered Alternatives involve changing this but the proposed
+  design does not.
+- The _upper_ of all of these shards are advanced in lockstep: whenever a table
+  write happens OR if it has been `X` time since the last round of upper
+  advancements. This allow joins of tables to be immediately served after a
+  write and joins of tables and sources to be served within `X`.
+- Timestamps for reads and writes are selected by an in-process _timestamp
```

Comment:
If we want to be very accurate: The oracle and the write lock/mutex are separate but they interact. The lock "protects" writes, not access to the timestamp oracle. A read-then-write transaction does:

1. acquire write lock (this prevents other writes from going through, which yes, indirectly also means that the oracle read timestamp will not advance)
2. get a read timestamp from the oracle
3. perform read
4. perform write (which also advances the read timestamp that the oracle keeps)
5. release write lock

Or I don't understand what _surrounding_ means here. 😅

Reply:
> which yes, indirectly also means that the oracle read timestamp will not advance

If I'm remembering correctly I don't think this is 100% accurate. The read timestamp will still advance from writes to system tables or from periodic table advancements, which both don't require the global write lock.

Reply:
Updated this with two bullets provided offline by joe.